File: src/main/GO/leetcode/lists/offer/Offer33.go
package offer

/**
 * @author wujingxinit@outlook.com
 * @date 2023/1/3 19:05
 */

func verifyPostorder(postorder []int) bool {

	if len(postorder) == 0 {
		return true
	}

	var dfs func(postorder []int, start int, end int) bool
	dfs = func(postorder []int, start int, end int) bool {
		if start >= end {
			return true
		}
		left := start
		for left < end && postorder[left] < postorder[end] {
			left++
		}
		right := left
		for right < end && postorder[right] > postorder[end] {
			right++
		}
		return right == end && dfs(postorder, start, left-1) && dfs(postorder, left, end-1)
	}
	return dfs(postorder, 0, len(postorder)-1)
}

File: java/src/leetcode/lists/offer/Offer33.java
package leetcode.lists.offer;

/**
 * @author wujingxinit@outlook.com
 * @date 2023/9/13 01:49
 */
public class Offer33 {
    public boolean verifyPostorder(int[] postorder) {
        return check(postorder, 0, postorder.length - 1);
    }

    public boolean check(int[] postorder, int start, int end) {
        if (start > end) {
            return true;
        }
        if (start == end) {
            return true;
        }
        int root = postorder[end];
        int index = end - 1;
        while (index > start && postorder[index] > root) {
            index--;
        }
        if (index > start) {
            // 剩下的都要比root小
            for (int i = start; i <= index; i++) {
                if (postorder[i] > root) {
                    return false;
                }
            }
        }
        return check(postorder, start, index) && check(postorder, index + 1, end - 1);
    }
}
